File: leetcode/1-two-sum.py
# 1-two-sum.py
#
# Copyright (C) 2019 Sang-Kil Park <likejazz@gmail.com>
# All rights reserved.
#
# This software may be modified and distributed under the terms
# of the BSD license.  See the LICENSE file for details.
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i, n in enumerate(nums):
            complement = target - n
            # Searches are much more faster than brute-force comparisons.
            if complement in nums[i + 1:]:
                return nums.index(n), nums[i + 1:].index(complement) + (i + 1)

File: go/1-two-sum.go
/* 1-two-sum.go
 *
 * Copyright (C) 2019 Sang-Kil Park <likejazz@gmail.com>
 * All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the BSD license.  See the LICENSE file for details.
 */
package solution

func twoSum(nums []int, target int) []int {
	numsMap := make(map[int]int)

	for i, num := range nums {
		numsMap[num] = i
	}

	for i, num := range nums {
		// Find the key immediately rather than the whole search.
		if j, ok := numsMap[target-num]; ok && i != j {
			return []int{i, j}
		}
	}
	// Prevent error, it doesn't mean anything.
	return []int{}
}
